### ISSUE TITLE
Invoke-DbaQuery doesn't use credential if DbaInstance used[Bug]

### ISSUE THREAD

**Issue by cfradenburg** [sqlcollaborative/dbatools#6008]
### Environmental information

```
#### PowerShell version:

Name                           Value                                                                                                                                             
----                           -----                                                                                                                                             
PSVersion                      5.1.17134.858                                                                                                                                     
PSEdition                      Desktop                                                                                                                                           
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                                                           
BuildVersion                   10.0.17134.858                                                                                                                                    
CLRVersion                     4.0.30319.42000                                                                                                                                   
WSManStackVersion              3.0                                                                                                                                               
PSRemotingProtocolVersion      2.3                                                                                                                                               
SerializationVersion           1.1.0.1                                                                                                                                           




#### dbatools Module version:


Name    : dbatools
Path    : C:\Program Files\WindowsPowerShell\Modules\dbatools\1.0.38\dbatools.psd1
Version : 1.0.38




```

```powershell
# Replace with output of Get-Culture
```

#### SQL Server: 

```sql
Microsoft SQL Server 2016 (SP2-GDR) (KB4505220) - 13.0.5101.9 (X64)
        Jun 15 2019 23:15:58
        Copyright (c) Microsoft Corporation
        Standard Edition (64-bit) on Windows Server 2016 Standard 10.0 <X64> (Build 14393: ) (Hypervisor)
```

```sql
/* REPLACE WITH output of @@LANGUAGE */
```

#### Host used

- [X] powershell.exe
- [ ] ISE
- [X] VS Code
- [ ] Other (please specify)

If anything other than powershell.exe was used, please confirm that you can duplicate the issue with powershell.exe

- [ ] Still buggy in powershell.exe

#### Steps to Reproduce

PS C:\> whoami
domain1\someuser
PS C:\> $domain2Cred = Get-Credential
cmdlet Get-Credential at command pipeline position 1
Supply values for the following parameters:
User: domain2\anotheruser
Password for user domain2\anotheruser: ***************
PS C:\> $serverConn = Connect-DbaInstance -SqlInstance servername.domain2.org -SqlCredential $domain2Cred

PS C:\> Invoke-DbaQuery -SqlInstance $serverConn -Database DBName -Query 'select suser_name()'

Column1
-------
domain1\someuser

#### Expected Behavior

That "domain2\anotheruser" would be returned

#### Actual Behavior

"domain1\someuser" was returned.  This only happens when passing in a database.  When excluding that parameter the correct user is returned.  Also, when passing in an instance name and credential the correct user is returned.

**Comment by cfradenburg** [comment]
I took a look at this and the `GetDatabaseConnection` method on lines 281 and 315 drop the credential.  I think replacing that with the following will fix it but I won't have time to test it today.

```$conncontext = New-Object Microsoft.SqlServer.Management.Smo.Server($server.InputObject)
$conncontext = .ConnectionContext.DatabaseName = $db.Name```

I'll pick this back up on Monday.

**Comment by wsmelton** [comment]
That method doesn't drop it, the purpose of that is to get the connection string with the database included.

We had to implement this for some reason around Azure if I recall, but can remember.

Any change made has to ensure this function still works against Azure SQL...just FYI.

**Comment by cfradenburg** [comment]
If I call `$server.ConnectionContext.Copy()` the ConnectAsUsername is preserved (at least in ad hoc code, it appears to drop it if I call it inside the function) but doing `$server.ConnectionContext.Copy().GetDatabaseConnection($Database)` removes that.  I follow the intent but I think there's an unwanted side effect there.

Everything I've tried either doesn't allow the database name to be changed, breaks the function, or doesn't carry the user through.  Thinking about it more it may be better to fix this in `Connect-DbaInstance` since it's not connecting using the specified database which is the reason for the if block starting on line 312.  It looks like in `Connect-DbaInstance` if you're passing in a server object it ignores the bound parameters due to the continues here:
```
            if ($instance.Type -like "Server" -or ($isAzure -and $instance.InputObject.ConnectionContext.IsOpen)) {
                if ($instance.InputObject.ConnectionContext.IsOpen -eq $false) {
                    $instance.InputObject.ConnectionContext.Connect()
                }
                if ($SqlConnectionOnly) {
                    $instance.InputObject.ConnectionContext.SqlConnectionObject
                    continue
                } else {
                    $instance.InputObject
                    [Sqlcollaborative.Dbatools.TabExpansion.TabExpansionHost]::SetInstance($instance.FullSmoName.ToLowerInvariant(), $instance.InputObject.ConnectionContext.Copy(), ($instance.InputObject.ConnectionContext.FixedServerRoles -match "SysAdmin"))

                    # Update cache for instance names
                    if ([Sqlcollaborative.Dbatools.TabExpansion.TabExpansionHost]::Cache["sqlinstance"] -notcontains $instance.FullSmoName.ToLowerInvariant()) {
                        [Sqlcollaborative.Dbatools.TabExpansion.TabExpansionHost]::Cache["sqlinstance"] += $instance.FullSmoName.ToLowerInvariant()
                    }
                    continue
                }
            }
```

However, I'm hesitant to remove those because that'll have a large impact.

**Comment by potatoqualitee** [comment]
Still valid

**Comment by jrobiii** [comment]
Don't know if this helps.  For me I've noticed this behavior in PS 7.0, but not in PS 5.1
```powershell
###############################################################################
# PS 5.1
###############################################################################
C:\work\Issues\> Get-Module dbatools

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     1.0.113    dbatools                            {Select-DbaObject, Set-DbatoolsConfig, Add-DbaAgDatabase, ...


C:\work\Issues\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.17763.1432
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.17763.1432
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1

C:\work\Issues\> $Credential = Get-Credential "<DOMAIN>\<SA_User>"
C:\work\Issues\> Invoke-DbaQuery -SqlInstance <SQLSERVER> -Query 'select suser_name()' -SqlCredential $Credential

Column1
-------
<DOMAIN>\<SA_User>

###############################################################################
# PS 7.0
###############################################################################
C:\work\Issues\> Get-Module dbatools

ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     1.0.113    dbatools                            {Select-DbaObject, Set-DbatoolsConfig, Add-DbaAgDatabase, ...


C:\work\Issues\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.0.3
PSEdition                      Core
GitCommitId                    7.0.3
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

C:\work\Issues\> $Credential = Get-Credential "<DOMAIN>\<SA_User>"
C:\work\Issues\> Invoke-DbaQuery -SqlInstance <SQLSERVER> -Query 'select suser_name()' -SqlCredential $Credential

Column1
-------
<DOMAIN>\<NONSA_User>

```

In 7.0 I get the user that I'm logged in as (NONSA_User) and not the user from $Credential (SA_User) and in 5.1 no such trouble.

**Comment by niphlod** [comment]
well, can you try with 

$Credential = Get-Credential "<DOMAIN>\<SA_User>"
$server = Connect-DbaInstance -Sqlinstance .... -Credential $Credential
$server.Query("select suser_name()")

?

**Comment by jrobiii** [comment]
Sorry, got kind of tied up.  Get-Credential without the domain fails with "Login failed for '<SQ_User>'.  With the domain in the credential the `$server.Query("select suser_name()")` still returns <DOMAIN>\<NONSA_User>

**Comment by andreasjordan** [comment]
Can you please try Connect-DbaInstance with the new code path (example 17 at https://docs.dbatools.io/#Connect-DbaInstance):

```
Set-DbatoolsConfig -FullName sql.connection.experimental -Value $true
$sqlcred = Get-Credential sqladmin
$server = Connect-DbaInstance -SqlInstance sql2014 -SqlCredential $sqlcred
Invoke-DbaQuery -SqlInstance $server -Query "select suser_name() as suser"
```

**Comment by andreasjordan** [comment]
I will close this for now, but you can reopen it if you have time to work on this.

**Comment by jrobiii** [comment]
Sorry that I didn't get back to @andreasjordan.  

I tried it again with Example 17.  With only the username (no domain) supplied to the `Get-Credential` I received "Login failed for <SA_User>", but when I supplied both the domain and user parts it worked.

I'll check with the rest of my team on the feasibility of using `Set-DbatoolsConfig -FullName sql.connection.experimental -Value $true` in production and non-production environments.

Thanks for the workaround.
